contributors/design-proposals/instrumentation/external-metrics-api.md

```diff
@@ -0,0 +1,87 @@
+
+
+# API
+
+The API will consist of a single path:
+
+
+```
+/apis/external.metrics.k8s.io/v1beta1/namespaces/<namespace_name>/<metric_name>?labelSelector=<selector>
+```
+
+Similar to endpoints in Custom Metrics API it would only support GET requests.
+
+The query would return the `ExternalMetricValueList` type described below:
+
+```go
+// a list of values for a given metric for some set labels
+type ExternalMetricValueList struct {
+       metav1.TypeMeta `json:",inline"`
+       metav1.ListMeta `json:"metadata,omitempty"`
+
+       // value of the metric matching a given set of labels
+       Items []ExternalMetricValue `json:"items"`
+}
+
+// a metric value for external metric
+type ExternalMetricValue struct {
+    metav1.TypeMeta`json:",inline"`
+
+    // the name of the metric
+    MetricName string `json:"metricName"`
+
+    // label set identifying the value within metric
+    MetricLabels map[string]string `json:"metricLabels"`
+
+    // indicates the time at which the metrics were produced
+    Timestamp unversioned.Time `json:"timestamp"`
+
+    // indicates the window ([Timestamp-Window, Timestamp]) from
+    // which these metrics were calculated, when returning rate
+    // metrics calculated from cumulative metrics (or zero for
+    // non-calculated instantaneous metrics).
+    WindowSeconds *int64 `json:"window,omitempty"`
+
+    // the value of the metric
+    Value resource.Quantity
+}
+```
+
+# Semantics
+
+## Namespaces
+
+Kubernetes namespaces don't have a natural 1-1 mapping to metrics coming from outside of Kubernetes. It is up to adapter implementing the API to decide which metric is available in which namespace. In particular a single metric may be available through many different namespaces.
+
+## Metric Values
+
+A request for a given metric may return multiple values if MetricSelector matches multiple time series. Each value should include a complete set of labels, which is sufficient to uniquely identify a timeseries.
+
+A single value should always be returned if MetricSelector specifies a single value for every label defined for a given metric.
+
+## Metric names
+
+Custom Metrics API [doesn't allow using certain characters in metric names](https://github.com/kubernetes/community/blob/master/contributors/design-proposals/instrumentation/custom-metrics-api.md#metric-names). The reason for that is a technical limitation in GO libraries. This list of forbidden characters includes slash (`/`). This is problematic as many systems use slashes in their metric naming convention.
+
+Rather than expect metric adapters to come up with their custom ways of handling that this document proposes introducing `\|` as a custom escape sequence for slash. HPA controller will automatically replace any slashes in MetricName field for External metric with this escape sequence.
+
+Otherwise the allowed metric names are the same as in Custom Metrics API.
+
+## Access Control
```

Comment:
A clear idea on access control (possibly with the mentioned ExternalMetricsPolicy) is definitely a GA requirement in my opinion. To get off the ground I think we can evaluate what makes sense once this lands.

Reply:
Agreed. I left it out of initial version because adding a policy would be a backward compatible change and I'd like to get some feedback from beta before committing to it.